### ISSUE TITLE
Can't use cache_nodes output of aws_elasticache_cluster as a list

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#826]
_This issue was originally opened by @phoolish as hashicorp/terraform#15040. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

This seems to be the return of hashicorp/terraform#8794

### Terraform Version
0.9.6
0.9.2

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_elasticache_cluster

### Terraform Configuration Files
```hcl
resource "aws_elasticache_cluster" "memcached" {
  cluster_id = "foo"
  engine = "memcached"

  num_cache_nodes = 3
  node_type = "cache.t2.micro"
}

resource "route_53_record" "memcached-site-com" {
  count = "${length(aws_elasticache_cluster.memcached.cache_nodes)}"
  zone_id = "${aws_route53_zone.site-com.zone_id}"
  name = "${format("node%d.mem.site.com", count.index)}"
  type = "CNAME"
  records = [
    "${element(aws_elasticache_cluster.memcached.cache_nodes.*.address, count.index)}",
  ]
  ttl = "300"
}
```

### Expected Behavior
3 route53_records should be generated.

### Actual Behavior
```
Error refreshing state: 1 error(s) occurred:

* aws_route53_record.mem-site-com: 3 error(s) occurred:

* aws_route53_record.mem-site-com[0]: At column 3, line 1: element: argument 1 should be type list, got type string in:

${element(aws_elasticache_cluster.mem-site-com.cache_nodes.*.address, count.index)}
* aws_route53_record.mem-site-com[1]: At column 3, line 1: element: argument 1 should be type list, got type string in:

${element(aws_elasticache_cluster.mem-site-com.cache_nodes.*.address, count.index)}

${element(aws_elasticache_cluster.mem-site-com.cache_nodes.*.address, 0)}
* aws_route53_record.mem-site-com[2]: At column 3, line 1: element: argument 1 should be type list, got type string in:

${element(aws_elasticache_cluster.mem-site-com.cache_nodes.*.address, count.index)}
```

### Steps to Reproduce
1. `terraform apply`

### References
- GH-10657

**Comment by Mattias-** [comment]
I'm getting this as well with 0.9.8. Would be nice with a fix.

**Comment by crielly** [comment]
Also running into this on 0.9.8, exact same use case :sob:

**Comment by imauriciovargas** [comment]
Running into the same issue creating CloudWatch alarms with version 0.9.11.

I get the following erros applying it all together: 

```
Error applying plan:

6 error(s) occurred:

* module.memcached.aws_cloudwatch_metric_alarm.evictions[0]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_
nodes.*.id'
* module.memcached.aws_cloudwatch_metric_alarm.evictions[1]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_
nodes.*.id'
* module.memcached.aws_cloudwatch_metric_alarm.CPU[1]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_nodes.
*.id'
* module.memcached.aws_cloudwatch_metric_alarm.swap[0]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_nodes
.*.id'
* module.memcached.aws_cloudwatch_metric_alarm.CPU[0]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_nodes.
*.id'
* module.memcached.aws_cloudwatch_metric_alarm.swap[1]: Resource 'aws_elasticache_cluster.memcached' does not have attribute 'cache_nodes.*.id' for variable 'aws_elasticache_cluster.memcached.cache_nodes
.*.id'
```

When I re-run terraform apply I get this error: 

```

**Comment by ccgagnon** [comment]
I got the same issue with 0.10.6.

**Comment by ssmereka** [comment]
Same issue in 0.10.8

**Comment by nshenry03** [comment]
```

**Comment by merlineus** [comment]
Same problem with. Trying to register cluster nodes in consul catalog:
 address = "${element(aws_elasticache_cluster.memcached.cache_nodes.*.address, count.index)}"

**Comment by ajbrown** [comment]
Any workarounds for this?

**Comment by imauriciovargas** [comment]
Yes. I found something today looking at a [reference](https://github.com/terraform-providers/terraform-provider-aws/issues/49) in the [bug](https://github.com/terraform-providers/terraform-provider-aws/issues/672). 

use [lookup](https://www.terraform.io/docs/configuration/interpolation.html#lookup-map-key-default-)

```
variable "num_nodes" {
  defaul  = 2
}

resource "aws_elasticache_cluster" "memcached" {
  cluster_id = "foo"
  engine = "memcached"
  num_cache_nodes = "${var.num_nodes}"
  ...
}

resource "aws_cloudwatch_metric_alarm" "CPU" {
  count                     = "${var.num_nodes}"
  ...
  dimensions                {
      CachedClusterId = "${aws_elasticache_cluster.memcached.id}",
      CachedNodeId    = "${lookup(aws_elasticache_cluster.memcached.cache_nodes[count.index], "id")}"
  }
}
```

Tested in Terraform version 10.6

**Comment by jdavisyei** [comment]
Is there a workaround that will let me dump the entire list of addresses into a comma separated list. 

I tried using various forms of `cache_nodes.*.address` but passing that to functions returns a count of the number of nodes instead of the address as I would expect. Using count.index unfortunately doesn't help because I need all of them in one object.

**Comment by dbutler-cb** [comment]
I was experiencing this same issue when trying to create an elasticache cluster using [this cloudposse wrapper module](https://github.com/cloudposse/terraform-null-label/) and Terraform v0.11.11. The result was a suppressed error message that results in:

`* module.this.aws_route53_record.livequery_elasticache: Resource 'aws_elasticache_cluster.this' not found for variable 'aws_elasticache_cluster.this.cache_nodes.0.address'`

Once I removed usage of that wrapper module it uncovered the actual error message and root cause which was the cluster I was attempting to create was >20 characters in length. Error message once I ran plan without the wrapper module:

`Error: module.livequery.aws_elasticache_cluster.livequery: "cluster_id" ("this-cluster-name-is-too-damn-long") must contain from 1 to 20 alphanumeric characters or hyphens`

**Comment by TechDawg270** [comment]
I was experiencing this same issue when trying to create an elasticache cluster using [this cloudposse wrapper module](https://github.com/cloudposse/terraform-null-label/) and Terraform v0.11.11. The result was a suppressed error message that results in:

`* module.this.aws_route53_record.this_elasticache: Resource 'aws_elasticache_cluster.this' not found for variable 'aws_elasticache_cluster.this.cache_nodes.0.address'`

Once I removed usage of that wrapper module it uncovered the actual error message and root cause which was the cluster I was attempting to create was >20 characters in length. Error message once I ran plan without the wrapper module:

`Error: module.livequery.aws_elasticache_cluster.this: "cluster_id" ("this-cluster-name-is-too-damn-long") must contain from 1 to 20 alphanumeric characters or hyphens`

**Comment by bflad** [comment]
The issue found by @TechDawg270 appears to be similar to https://github.com/hashicorp/terraform/issues/18129 which is fixed in Terraform 0.12. 👍

**Comment by bflad** [comment]
Hi folks 👋 This issue is resolved in Terraform 0.12, which fully supports indexed splat (`*`) notation with list attribute values.

Given this configuration:

```hcl
terraform {
  required_providers {
    aws = "2.20.0"
  }
  required_version = "0.12.5"
}

provider "aws" {
  region = "us-east-2"
}

resource "aws_elasticache_cluster" "test" {
  cluster_id      = "bflad-testing"
  engine          = "memcached"
  node_type       = "cache.t2.micro"
  num_cache_nodes = 3
}

output "test1" {
  value = aws_elasticache_cluster.test.cache_nodes[*].address
}

output "test2" {
  value = element(aws_elasticache_cluster.test.cache_nodes[*].address, 1)
}

output "test3" {
  value = join(",", formatlist("%s:%s", aws_elasticache_cluster.test.cache_nodes[*].address, aws_elasticache_cluster.test.cache_nodes[*].port))
}
```

We can get the various expected outputs that were problematic in Terraform 0.11 and earlier:

```
Outputs:

test1 = [
  "bflad-testing.asxu6y.0001.use2.cache.amazonaws.com",
  "bflad-testing.asxu6y.0002.use2.cache.amazonaws.com",
  "bflad-testing.asxu6y.0003.use2.cache.amazonaws.com",
]
test2 = bflad-testing.asxu6y.0002.use2.cache.amazonaws.com
test3 = bflad-testing.asxu6y.0001.use2.cache.amazonaws.com:11211,bflad-testing.asxu6y.0002.use2.cache.amazonaws.com:11211,bflad-testing.asxu6y.0003.use2.cache.amazonaws.com:11211
```

Enjoy! 🚀
